raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2408,19 +2418,32 @@ private void fireHandleCommit(BatchReader<T> reader) {
             listener.handleCommit(reader);
         }
 
-        void maybeFireHandleClaim(int epoch, long epochStartOffset) {
-            // We can fire `handleClaim` as soon as the listener has caught
-            // up to the start of the leader epoch. This guarantees that the
-            // state machine has seen the full committed state before it becomes
-            // leader and begins writing to the log.
-            if (epoch > claimedEpoch && nextOffset() >= epochStartOffset) {
-                claimedEpoch = epoch;
-                listener.handleClaim(epoch);
+        void maybeFireLeaderChange(LeaderAndEpoch leaderAndEpoch) {
+            if (shouldFireLeaderChange(leaderAndEpoch)) {
+                lastFiredLeaderChange = leaderAndEpoch;
+                listener.handleLeaderChange(leaderAndEpoch);
             }
         }
 
-        void fireHandleResign(int epoch) {
-            listener.handleResign(epoch);
+        private boolean shouldFireLeaderChange(LeaderAndEpoch leaderAndEpoch) {
+            if (leaderAndEpoch.equals(lastFiredLeaderChange)) {
+                return false;
+            } else if (leaderAndEpoch.epoch > lastFiredLeaderChange.epoch) {
+                return true;
+            } else {
+                return leaderAndEpoch.leaderId.isPresent() && !lastFiredLeaderChange.leaderId.isPresent();
```

Comment:
This works because there is an invariant that `leaderAndEpoch.epoch >= lastFiredLeaderChange.epoch` is always `true`, right? Should we document this above this line?

Thinking about it some more, wouldn't this always be true since we know that:
1. `leaderAndEpoch.epoch == lastFiredLeaderChange.epoch`
2. `!leaderAndEpoch.equals(lastFiredLeaderChange)`

If you agree, I think that we can change this method to just:
```
return !leaderAndEpoch.equals(lastFiredLeaderChange);
```